mayastor/src/host/device.rs

```diff
@@ -0,0 +1,327 @@
+//!
+//! This module implements the list_block_devices() gRPC method
+//! for listing available disk devices on the current host.
+//!
+//! The relevant information is obtained via udev.
+//! The method works by iterating through udev records and selecting block
+//! (ie. SUBSYSTEM=block) devices that represent either disks or disk
+//! partitions. For each such device, it is then determined as to whether the
+//! device is available for use.
+//!
+//! A device is currently deemed to be "available" if it satisfies the following
+//! criteria:
+//!  - the device has a non-zero size
+//!  - the device is of an acceptable type as determined by well known device
+//!    numbers (eg. SCSI disks)
+//!  - the device represents either a disk with no partitions or a disk
+//!    partition of an acceptable type (Linux filesystem partitions only at
+//!    present)
+//!  - the device currently contains no filesystem or volume id (although this
+//!    logically implies that the device is not currently mounted, for the sake
+//!    of consistency, the mount table is also checked to ENSURE that the device
+//!    is not mounted)
+
+use std::{
+    collections::HashMap,
+    ffi::{OsStr, OsString},
+    io::Error,
+};
+
+use proc_mounts::{MountInfo, MountIter};
+use rpc::mayastor::{
+    block_device::{Filesystem, Partition},
+    BlockDevice,
+};
+use udev::{Device, Enumerator};
+
+// Struct representing a property value in a udev::Device struct (and possibly
+// elsewhere). It is used to provide conversions via various "From" trait
+// implementations below.
+struct Property<'a>(Option<&'a OsStr>);
+
+impl From<Property<'_>> for String {
+    fn from(property: Property) -> Self {
+        String::from(property.0.map(|s| s.to_str()).flatten().unwrap_or(""))
+    }
+}
+
+impl From<Property<'_>> for Option<String> {
+    fn from(property: Property) -> Self {
+        property.0.map(|s| s.to_str()).flatten().map(String::from)
+    }
+}
+
+impl From<Property<'_>> for Option<u32> {
+    fn from(property: Property) -> Self {
+        Option::<String>::from(property)
+            .map(|s| s.parse().ok())
+            .flatten()
+    }
+}
+
+impl From<Property<'_>> for u32 {
+    fn from(property: Property) -> Self {
+        Option::<Self>::from(property).unwrap_or(0)
+    }
+}
+
+impl From<Property<'_>> for Option<u64> {
+    fn from(property: Property) -> Self {
+        Option::<String>::from(property)
+            .map(|s| s.parse().ok())
+            .flatten()
+    }
+}
+
+impl From<Property<'_>> for u64 {
+    fn from(property: Property) -> Self {
+        Option::<Self>::from(property).unwrap_or(0)
+    }
+}
+
+// Determine the type of devices which may be potentially presented
+// as "available" for use.
+fn usable_device(devmajor: &u32) -> bool {
```

Comment:
what is the major for nvme? Also, have you considered leveraging https://github.com/stratis-storage/libblkid-rs it is we already have liblkid iirc. Not sure if it brings anything for our use case right now, perhaps something for the future?

Reply:
nvme ends up with major 259.
Will investigate libblkid-rs as possible future enhancement, although the goal here was to keep it as simpe as possible at least in the first instance.